### PR TITLE
(application) Updated with customtkinter; Volume Controls and major-minor improvements

### PR DESCRIPTION
**Pull request by Sayad-Uddin-Tahsin**
- Added Volume Mute on Session Break (Windows Only)
- Added Fade in and out transition to sound (Windows Only)
- Fixed `Skip this time` timer issue-
  Previously in the case of skip the remaining 20 seconds used to add with the next 20 minute on interval.
- Removed `blink-eye-reminder-btn.png` and used Button instead with 2 skip images (for `dark` and `light` mode)
- Changed fonts on Break Screen
- Changed the spacing on the links at the bottom of break screen
- Break screen now synced with System Theme (`dark` and `light`) as well as the widgets color on it
- Modified links to once clickable in one break screen
- Added `hold_the_program` instead of `time.sleep` to wait the program in order to exclude remaining time in the case of `skip`
- Removed `--icon` for Fedora, as https://github.com/nomandhoni-cs/blink-eye/pull/29#pullrequestreview-1975546330
- Added `logo` and text `Blink Eye` on the top right side in the Break Screen

@fluentmoheshwar Please consider reviewing Fedora related files (especially, [./application/BlinkEye_Fedora.spec](https://github.com/nomandhoni-cs/blink-eye/blob/master/application/BlinkEye_Fedora.spec)) once again, as some lines were deleted while merging #26!

### CODE REVIEW

**Comment by fluentmoheshwar** [comment]
Please don't merge until I review it. I have to wait until the electricity comes back to test it.

**Comment by fluentmoheshwar** [comment]
Anyway, I have read the code with my phone. Can we add multiple fonts? As we don't have Segoe UI in fedora.

**Comment by Sayad-Uddin-Tahsin** [comment]
> Anyway, I have read the code with my phone. Can we add multiple fonts? As we don't have Segoe UI in fedora.

What font you're trying to mean?

**Comment by Sayad-Uddin-Tahsin** [comment]
> > Anyway, I have read the code with my phone. Can we add multiple fonts? As we don't have Segoe UI in fedora.
> 
> What font you're trying to mean?

Do we have Helvetica on Fedora? Segoe UI seems really cool to see so added, it's changeable obviously!

**Comment by fluentmoheshwar** [comment]
> > > Anyway, I have read the code with my phone. Can we add multiple fonts? As we don't have Segoe UI in fedora.
> > 
> > What font you're trying to mean?
> 
> Do we have Helvetica on Fedora? Segoe UI seems really cool to see so added, it's changeable obviously! 

Helvatica isn't pre-installed on Fedora. The ideal solution would be

- Segoe UI on Windows
- Cantarell on Fedora/Debian/Opensuse and others
- Ubuntu on Ubuntu
- SF Pro on macOS

**Comment by Sayad-Uddin-Tahsin** [comment]
> > > > Anyway, I have read the code with my phone. Can we add multiple fonts? As we don't have Segoe UI in fedora.
> > > 
> > > 
> > > What font you're trying to mean?
> > 
> > 
> > Do we have Helvetica on Fedora? Segoe UI seems really cool to see so added, it's changeable obviously!
> 
> Helvatica isn't pre-installed on Fedora. The ideal solution would be
> 
> * Segoe UI on Windows
> * Cantarell on Fedora/Debian/Opensuse and others
> * Ubuntu on Ubuntu
> * SF Pro on macOS

In that case use of which would be ideal? Helvetica was used before!

**Comment by fluentmoheshwar** [comment]
> > > > > Anyway, I have read the code with my phone. Can we add multiple fonts? As we don't have Segoe UI in fedora.
> > > > 
> > > > 
> > > > What font you're trying to mean?
> > > 
> > > 
> > > Do we have Helvetica on Fedora? Segoe UI seems really cool to see so added, it's changeable obviously!
> > 
> > Helvatica isn't pre-installed on Fedora. The ideal solution would be
> > 
> > * Segoe UI on Windows
> > * Cantarell on Fedora/Debian/Opensuse and others
> > * Ubuntu on Ubuntu
> > * SF Pro on macOS
> 
> In that case use of which would be ideal? Helvetica was used before! 

is there any way to add fallback fonts like we do in CSS? I never really worked with tkinter.

**Comment by Sayad-Uddin-Tahsin** [comment]
> > > > > > > > Anyway, I have read the code with my phone. Can we add multiple fonts? As we don't have Segoe UI in fedora.
> > > > > > > 
> > > > > > > 
> > > > > > > What font you're trying to mean?
> > > > > > 
> > > > > > 
> > > > > > Do we have Helvetica on Fedora? Segoe UI seems really cool to see so added, it's changeable obviously!
> > > > > 
> > > > > 
> > > > > Helvatica isn't pre-installed on Fedora. The ideal solution would be
> > > > > 
> > > > > * Segoe UI on Windows
> > > > > * Cantarell on Fedora/Debian/Opensuse and others
> > > > > * Ubuntu on Ubuntu
> > > > > * SF Pro on macOS
> > > > 
> > > > 
> > > > In that case use of which would be ideal? Helvetica was used before!
> > > 
> > > 
> > > is there any way to add fallback fonts like we do in CSS? I never really worked with tkinter.
> > 
> > 
> > Maybe we can import font from `.ttf` file! I'll commit the changes with the modification later. Can you just review Fedora things, especially the spec file?
> 
> Okay! I have reviewed it. But I want to run the script and test but I don't have electricity right now. I will test and review immediately after the electricity come back. Anyway if you want to use ".ttf" files don't use something like Segoe UI. It will cause licensing issues. Use a font that's open source like Noto Sans or Ubuntu.

Roger that! 🫡

**Comment by fluentmoheshwar** [comment]
Build script needs to be changed and comtypes isn't available on linux

**Comment by Sayad-Uddin-Tahsin** [comment]
> Build script needs to be changed and comtypes isn't available on linux

In that case, volume feature might not be available for Linux! Is there any way you know which is OS Independent? I can't find any actually...

**Comment by Sayad-Uddin-Tahsin** [comment]
> Update `BlinkEye_Fedora.spec` to this:
> 
> ```python
> Name:           BlinkEye
> Version:        1.4.0
> Release:        1%{?dist}
> Summary:        Blink Eye is a minimalist eye care reminder app designed to reduce eye strain during extended screen usage.
> 
> License:        MIT
> URL:            https://github.com/nomandhoni-cs/blink-eye
> 
> %description
> Blink Eye is a minimalist eye care reminder app designed to reduce eye strain during extended screen usage. It provides customization timers, full-screen popups, audio mute functionality for a seamless user experience.
> 
> %install
> rm -rf $RPM_BUILD_ROOT
> mkdir -p $RPM_BUILD_ROOT/%{_bindir}
> cp dist/%{name} $RPM_BUILD_ROOT/%{_bindir}
> 
> %clean
> rm -rf $RPM_BUILD_ROOT
> 
> %files
> %{_bindir}/%{name}
> %license LICENSE.txt
> 
> %changelog
> * Mon Apr 01 2024 Moheshwar Amarnath Biswas
> - First Linux Version
> ```
> 
> and `build_fedora.sh` to this:
> 
> ```shell
> #!/usr/bin/env bash
> 
> ver="1.4.0"
> 
> echo Installing Dependencies...
> sudo dnf install -y rpmdevtools rpmlint python3-devel python3-pip python3-tkinter
> pip install wheel
> pip install -r REQUIREMENTS.txt
> pip install pyinstaller
> echo Building the executable...
> pyinstaller --name BlinkEye --onefile --windowed --add-data="./Assets/*:./Assets" --hidden-import plyer.platforms.linux.notification --clean blink-eye.py
> echo Building the installer...
> rpmdev-setuptree
> cp -r ./* ~/rpmbuild/BUILD
> cp ../LICENSE.txt ~/rpmbuild/BUILD/LICENSE.txt
> rpmbuild -bb ./BlinkEye_Fedora.spec
> rm -rf ~/rpmbuild/BUILD
> mkdir -p ./ExecutableFile/
> mv ~/rpmbuild/RPMS/x86_64/BlinkEye-$ver-1.fc39.x86_64.rpm ./ExecutableFile/BlinkEye-$ver-1.fc39.x86_64.rpm
> ```
> 
> and if else block to stop importing `comtypes` on macOS and Linux

The spec file won't be able to access the LINCENSE.txt as it's one step back in PATH, I think? Should I still edit it?

**Comment by fluentmoheshwar** [comment]
> > Update `BlinkEye_Fedora.spec` to this:
> > 
> > ```python
> > Name:           BlinkEye
> > Version:        1.4.0
> > Release:        1%{?dist}
> > Summary:        Blink Eye is a minimalist eye care reminder app designed to reduce eye strain during extended screen usage.
> > 
> > License:        MIT
> > URL:            https://github.com/nomandhoni-cs/blink-eye
> > 
> > %description
> > Blink Eye is a minimalist eye care reminder app designed to reduce eye strain during extended screen usage. It provides customization timers, full-screen popups, audio mute functionality for a seamless user experience.
> > 
> > %install
> > rm -rf $RPM_BUILD_ROOT
> > mkdir -p $RPM_BUILD_ROOT/%{_bindir}
> > cp dist/%{name} $RPM_BUILD_ROOT/%{_bindir}
> > 
> > %clean
> > rm -rf $RPM_BUILD_ROOT
> > 
> > %files
> > %{_bindir}/%{name}
> > %license LICENSE.txt
> > 
> > %changelog
> > * Mon Apr 01 2024 Moheshwar Amarnath Biswas
> > - First Linux Version
> > ```
> > 
> > and `build_fedora.sh` to this:
> > 
> > ```shell
> > #!/usr/bin/env bash
> > 
> > ver="1.4.0"
> > 
> > echo Installing Dependencies...
> > sudo dnf install -y rpmdevtools rpmlint python3-devel python3-pip python3-tkinter
> > pip install wheel
> > pip install -r REQUIREMENTS.txt
> > pip install pyinstaller
> > echo Building the executable...
> > pyinstaller --name BlinkEye --onefile --windowed --add-data="./Assets/*:./Assets" --hidden-import plyer.platforms.linux.notification --clean blink-eye.py
> > echo Building the installer...
> > rpmdev-setuptree
> > cp -r ./* ~/rpmbuild/BUILD
> > cp ../LICENSE.txt ~/rpmbuild/BUILD/LICENSE.txt
> > rpmbuild -bb ./BlinkEye_Fedora.spec
> > rm -rf ~/rpmbuild/BUILD
> > mkdir -p ./ExecutableFile/
> > mv ~/rpmbuild/RPMS/x86_64/BlinkEye-$ver-1.fc39.x86_64.rpm ./ExecutableFile/BlinkEye-$ver-1.fc39.x86_64.rpm
> > ```
> > 
> > and if else block to stop importing `comtypes` on macOS and Linux
> 
> The spec file won't be able to access the LINCENSE.txt as it's one step back in PATH, I think? Should I still edit it? 

The spec file isn't reading from current directory it's reading from `~/rpmbuild/BUILD` don't edit it

**Comment by fluentmoheshwar** [comment]
> > Build script needs to be changed and comtypes isn't available on linux
> 
> In that case, volume feature might not be available for Linux! Is there any way you know which is OS Independent? I can't find any actually... 

I can't find any either.

**Review by fluentmoheshwar — APPROVED** [review]
LGTM! You should add if else statement to comtype imports and merge it we can make that feature cross platform later

**Comment by Sayad-Uddin-Tahsin** [comment]
> LGTM! You should add if else statement to comtype imports and merge it we can make that feature cross platform later

Yeah thinking to do so, trying to figure out the exception type!

**Comment by fluentmoheshwar** [comment]
> > LGTM! You should add if else statement to comtype imports and merge it we can make that feature cross platform later
> 
> Yeah thinking to do so, trying to figure out the exception type! 

You should try something like this

```python
import platform
if platform.system == "Windows":
  # import comtypes and implement volume feature
```

**Comment by Sayad-Uddin-Tahsin** [comment]
@nomandhoni-cs Please review the Website change. And allow us to view the preview web, as it helps to see the changes and debug!

**Comment by fluentmoheshwar** [comment]
> @fluentmoheshwar, on the latest commit, tried to add font import mechanism but failed miserably. But added the font path everywhere the font to be used, so that it could easily be identifiable what are meant to be used. Currently it should use the default OS font.
> 
> I hope it's totally ok!

Let it use default os font

**Comment by Sayad-Uddin-Tahsin** [comment]
@fluentmoheshwar Done! It will import `NotoSans-Regular` and `Consolas` font on use!

**Comment by fluentmoheshwar** [comment]
> @fluentmoheshwar Done! It will import `NotoSans-Regular` and `Consolas` font on use! 

Awesome!

File: application/blink-eye.py
Comment (nomandhoni-cs) [inline application/blink-eye.py:40]:
```suggestion
ctk.FontManager.load_font(resource_path("NotoSans-Regular.ttf"))
```

File: application/blink-eye.py
Comment (nomandhoni-cs) [inline application/blink-eye.py:41]:
```suggestion
ctk.FontManager.load_font(resource_path("Consolas.ttf"))
```

**Review by nomandhoni-cs — CHANGES_REQUESTED** [review]

**Comment by nomandhoni-cs** [comment]
Screenshot_36

I have tested your changes.
This was the major problem, this wasn't creating a full screen reminder.
this was most probably because of `customtkinter`

# Issues with the new changes you made,
- Not creating a fullscreen popup
- Remove the app name and logo from the popup, our target is keep it as minimal as possible
- Change the new skip this time btn to match the previous theme

Here is our figma file, you can color code and icon and spacing from here

https://www.figma.com/file/8GenlMVGFGTpo2dLzVkyh0/Blink-Eye?type=design&node-id=21%3A7&mode=design&t=q3WCB2XnOM7sgfMC-1

**Comment by Sayad-Uddin-Tahsin** [comment]
> Screenshot_36
> 
> I have tested your changes. This was the major problem, this wasn't creating a full screen reminder. this was most probably because of `customtkinter`
> 
> # Issues with the new changes you made,
> * Not creating a fullscreen popup
> * Remove the app name and logo from the popup, our target is keep it as minimal as possible
> * Change the new skip this time btn to match the previous theme
> 
> Here is our figma file, you can get color code and icon and spacing from here
> 
> https://www.figma.com/file/8GenlMVGFGTpo2dLzVkyh0/Blink-Eye?type=design&node-id=21%3A7&mode=design&t=q3WCB2XnOM7sgfMC-1

• Full screen pop up will be fixed soon in new commit from my PC and it's totally not a customtkinter issue I think.
• In my opinion, on the break screen, there is no way to get which app is showing the popup. In order to aware the user of the software's existence and some sort of promote as someone sees the break screen, they might be curious and they'll get the name of the software automatically. 
• Using Image as Button, doesn't seems professional as we can create rounded buttons with hovering colors which can't be done in Image. Will it be ok, if I match the color and icon? And use a darkish color as `hover_color`?

**Comment by Sayad-Uddin-Tahsin** [comment]
> Screenshot_36
> 
> I have tested your changes. This was the major problem, this wasn't creating a full screen reminder. this was most probably because of `customtkinter`
> 
> # Issues with the new changes you made,
> * Not creating a fullscreen popup
> * Remove the app name and logo from the popup, our target is keep it as minimal as possible
> * Change the new skip this time btn to match the previous theme
> 
> Here is our figma file, you can get color code and icon and spacing from here
> 
> https://www.figma.com/file/8GenlMVGFGTpo2dLzVkyh0/Blink-Eye?type=design&node-id=21%3A7&mode=design&t=q3WCB2XnOM7sgfMC-1

I really don't know how and why it's happening, seems everything is perfect here: 
BlinkEye Screenshot
While running `.py` and `.exe`, both working perfectly! Can you just try running on different PC? or just try re-executing? if both fails, can you please run `.py` and see what happens?

**Comment by nomandhoni-cs** [comment]
Thanks a lot for all the efforts you put together.
And Special thanks for the volume control. Probably we will try to add a unmute button beside skip reminder btn

**Comment by Sayad-Uddin-Tahsin** [comment]
> Thanks a lot for all the efforts you put together. And Special thanks for the volume control. Probably we will try to add a unmute button beside skip reminder btn

Great Idea!

**Pull request merged**
